### PR TITLE
[Structural] Wrinkling claw 2d

### PR DESCRIPTION
**Pull request by KlausBSautter**
Hi, this sould be the last feature for the membrane element for now.
This PR adds a linear wrinkling claw and adjusts the membrane element itself.

The application of the claw is demonstrated in the following pictures. 

**Standard formulation: Wrinklies and unrealistic compression stresses** 
no_tfl

**Tension field method: Smooth surface with min stress = 0**
tfl

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+{
+    KRATOS_ERROR_IF(mpConstitutiveLaw->WorkingSpaceDimension()<2) << "WorkingSpaceDimension must be bigger than 1" << std::endl;
+    return mpConstitutiveLaw->WorkingSpaceDimension();
+}
+
+
+std::size_t WrinklingLinear2DLaw::GetStrainSize()
+{
+    SizeType strain_size = 3;
+    KRATOS_ERROR_IF_NOT(mpConstitutiveLaw->GetStrainSize()==3) << "wrinkling law only works for 2D base laws (strain size = 3)" << std::endl;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]:
```suggestion
    KRATOS_ERROR_IF_NOT(mpConstitutiveLaw->GetStrainSize()==3) << "Wrinkling law only works for 2D base laws (strain size = 3)" << std::endl;
```

**Comment by KlausBSautter** [comment]
this needs #6392 to be merged

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+
+    void InitializeNonLinearIteration(
+        const Properties& rMaterialProperties,
+        const GeometryType& rElementGeometry,
+        const Vector& rShapeFunctionsValues,
+        const ProcessInfo& rCurrentProcessInfo
+        ) override;
+
+
+    void FinalizeNonLinearIteration(
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h]: Can you please use only the new Interface of the CLaw?

Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h]: this is just a "wrapper" for any other law. So as long as they use the old functions I also have to provide them? I think :thinking:

Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h]: then we should update the other laws (I thought they are already)
it has been deprecated for a long time

Can you check please which law is not yet updated?

Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h]: @philbucher, of course, you were right. I removed all deprecated functions, they were unnecessary... :penguin:

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+    element_parameters.SetMaterialProperties(rValues.GetMaterialProperties());
+    element_parameters.SetStrainVector(strain_vector);
+    element_parameters.SetStressVector(stress_vector);
+    element_parameters.SetConstitutiveMatrix(material_tangent_modulus);
+    element_parameters.Set(ConstitutiveLaw::COMPUTE_STRESS);
+    element_parameters.Set(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN);
+    element_parameters.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR);
+    mpConstitutiveLaw->CalculateMaterialResponsePK2(element_parameters);
+
+    // consider initial pre-stress state which is calculated on element level (i.e. projection)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]: I would tsrta all comments with capital letter

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+        const double n_1 = wrinkling_direction[0];
+        const double n_2 = wrinkling_direction[1];
+
+
+        Vector wrinkling_operation_vector_1 = ZeroVector(3);
+        wrinkling_operation_vector_1[0] = n_1*n_1;
+        wrinkling_operation_vector_1[1] = n_2*n_2;
+        wrinkling_operation_vector_1[2] = n_1*n_2*2.0;
+
+        Vector temp_vec = prod(material_tangent_modulus,wrinkling_operation_vector_1);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]:
```suggestion
        const Vector temp_vec = prod(material_tangent_modulus,wrinkling_operation_vector_1);
```

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+        Matrix temp_mat = outer_prod(temp_vec,wrinkling_operation_vector_1);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]:
```suggestion
        const Matrix temp_mat = outer_prod(temp_vec,wrinkling_operation_vector_1);
```

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+        Matrix material_tangent_modulus_modified_1 = prod(temp_mat,material_tangent_modulus);
+        double temp_double = inner_prod(wrinkling_operation_vector_1,temp_vec);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]:
```suggestion
        const double temp_double = inner_prod(wrinkling_operation_vector_1,temp_vec);
```

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+        // pre-stress was only needed for wrinkling check
+        // pre-stress is added in the element
+        stress_vector -= pre_stress_vector;
+    }
+
+    // set the data on the main claw
+    Flags& r_constitutive_law_options = rValues.GetOptions();
+    if( r_constitutive_law_options.Is( ConstitutiveLaw::COMPUTE_STRESS )) {
+        Vector& r_stress_vector = rValues.GetStressVector();
+        r_stress_vector = stress_vector;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]:
```suggestion
        noalias(r_stress_vector) = stress_vector;
```

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+    }
+
+    if( r_constitutive_law_options.Is( ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR )) {
+        Matrix& r_constitutive_matrix = rValues.GetConstitutiveMatrix();
+        r_constitutive_matrix = material_tangent_modulus;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]:
```suggestion
        noalias(r_constitutive_matrix) = material_tangent_modulus;
```

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+     */
+    SizeType WorkingSpaceDimension() override;
+
+    /**
+     * @brief Voigt tensor size
+     * @details This is not used, so 0 is returned
+     */
+    SizeType GetStrainSize() override;
+
+
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h:124]: Docimentation is missing

**Review by loumalouomega — COMMENTED** [review]
Comments

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h (hunk 9 above)
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h:124]: documentation for virtual functions?

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h:124]: IMO

Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h:124]: alright, so just copy the documentation of the virtual function?

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h:124]: yes

**Comment by KlausBSautter** [comment]
any comments? @philbucher

**Comment by KlausBSautter** [comment]
@philbucher I moved the functions to the

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+
+ConstitutiveLaw::Pointer WrinklingLinear2DLaw::Create(Kratos::Parameters NewParameters) const
+{
+    return Kratos::make_shared<WrinklingLinear2DLaw>();
+}
+
+//*******************************DESTRUCTOR*******************************************
+/***********************************************************************************/
+
+WrinklingLinear2DLaw::~WrinklingLinear2DLaw()
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]: can be moved to the header

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Klaus B. Sautter
+//
+// System includes
+#include <iostream>
+#include <set>
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]:
```suggestion

```

I don't think you use this

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+//*******************************DESTRUCTOR*******************************************
+/***********************************************************************************/
+
+WrinklingLinear2DLaw::~WrinklingLinear2DLaw()
+{
+};
+
+
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]: there are many empty lines in this file, I would separate with one empty line btw functions

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h [hunk 13]
```diff
@@ -0,0 +1,615 @@
...
+    /**
+     * @brief Copy constructor.
+     */
+    WrinklingLinear2DLaw (const WrinklingLinear2DLaw& rOther);
+
+    /**
+     * @brief Destructor.
+     */
+    ~WrinklingLinear2DLaw() override;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.h]:
```suggestion
    ~WrinklingLinear2DLaw() override = default;
```

**Review by philbucher — COMMENTED** [review]
minor otherwise looks good

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp (hunk 12 above)
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]: >  I would separate with one empty line btw functions

?

Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/wrinkling_linear_2d_law.cpp]: you often have 3 or more empty lines btw functions
=> reduce to 1

**Review by philbucher — APPROVED** [review]

**Pull request merged**